In [73]:
"""Numeric Analysis with Numpy"""

'Numeric Analysis with Numpy'

In [74]:
 #!/usr/bin/env python

# Lesson 2

## Question 1

In [75]:
import numpy as np

1. Use the randn function to create an array with a dimension of 5X5 and use a for loop to calculate the sum of all elements in the diagonal of the array. (25 points)

To begin, I used the randn function to simulate data from the normal distribution.

In [76]:
np.random.seed(17)
x = np.random.randn(5, 5)


In [77]:
def diag_sum(arr):
    """Calculate the sum of all elements in the diagonal of an array."""
    result = 0
    for i in range(5):
        result += arr[i, i]
    return "The sum of elements in the diagonal of the array is: ", result

I applied the `diag_sum()` function on the simulated data, which returned a sum of -3.553.

In [78]:
diag_sum(x)

('The sum of elements in the diagonal of the array is: ', -3.5528347220435976)

2. Choose any three functions to apply to this array. (25 points)

In [79]:
print(x.min())

-1.854628078806505


The first function I applied to the array is `min()` to find the smallest value in the data. This returned a value of 1.855.

In [80]:
print(x.cumprod())

[ 2.76265890e-01 -5.12370477e-01 -3.19668510e-01 -3.66119953e-01
 -3.79736126e-01 -7.16424958e-01  8.00234428e-02 -2.89765957e-02
 -4.30809668e-03  1.88601215e-03  4.09501707e-03  4.71873014e-03
 -8.58248461e-03  1.18480634e-03  6.39605399e-04 -1.13548014e-03
 -1.49301619e-03  7.06865598e-04 -7.72059742e-04  1.93036122e-04
 -1.89618283e-04 -1.95547475e-04 -9.60790791e-05  4.29133944e-05
 -3.46036482e-05]


The next function I applied was `cumprod()`. This function calculated the cumulative product of the values. The resulting array can be seen above.

In [81]:
print(x.mean())

0.05078031173984357


The final function I applied to the array was `mean()`. The mean of all values in the data was 0.051.

## Question 2

1. Use x = np.random.randint(0, 1000, size = (10, 10)) to generate 10x10 array and use a for loop to find out how many even numbers are in it. (25 points)

For this simulated data, I used `randint` to generate 100 values between 1 and 1000.

In [82]:
np.random.seed(17)
y = np.random.randint(0, 1000, size = (10, 10))

In [83]:
def count_even_nums(arr):
    """Return the count of even numbers in an array."""
    even_nums = 0
    for i in arr:
        for j in i:
            if j%2==0:
                even_nums += 1
    return "There are", even_nums, "even numbers in the array."


I applied the `count_even_nums()` function on the simulated data, which returned a total of 42 even numbers.

In [84]:
count_even_nums(y)

('There are', 42, 'even numbers in the array.')

2. Randomly generate an 8x9 array from a normal distribution with mean = 1, sigma = 0.5. Calculate the mean of elements whose indexes have a relation of (i+j)%5 == 0  (i is row index and j is column index).

I used the `normal` function to simulate values from the N(1, 0.5) distribution.

In [85]:
z = np.random.normal(loc=1, scale=0.5, size=(8, 9))

In [86]:
def five_divisible_ind(z):
    """Calculate the mean of elements in an array whose indexes' sum are divisible by five."""
    result_arr = np.array([z[i, j] for i in range(8) for j in range(9) if (i+j)%5==0])
    return "For all elements in the array whose indexes sums is divisible by 5, their mean is: ", result_arr.mean()

I applied the `five_divisible_ind()` function on the simulated data, which returned a mean of 0.94.

In [87]:
five_divisible_ind(z)

('For all elements in the array whose indexes sums is divisible by 5, their mean is: ',
 0.9400481073729212)